design-documents/storefront/catalog/product-options-and-variants.md

```diff
@@ -0,0 +1,356 @@
+* Align a similar option with a single data structure to manage them together.
+
+
+## Modeling Options and Variants Data Objects
+
+The great advantage of Magento 2 - modularity was not a real thing for the product options during Magento the lifetime.
+Since times of Magento 1.x option prices coupled into options, inventory records ignored even options with assigned SKUs.
+The only thing that had to play it together "Bundle" product was designed too complex to treat it as universal solution it had to become.
+
+That's why, with the storefront project, we have an opportunity to resolve it.
+
+Taking into account the Magento experience (good, bad, ugly) and solutions proposed on the market.
+I could define the set of requirements that I would want to see in a new options implementation:
+
+* Options are a predefined list of possible product customizations. This list belongs to the product exclusively; it is measurable, and limited only by the business requirements. So it always easy to return even a complete set of options for rendering PDP initially. 
+
+* Variants are a unique intersection of one or may option values. The variant matrix does not belong to a product as property. The variant matrix is stored managed separately from products. The primary role of variants is to distinguish the possible intersections of options from impossible.
+The variants matrix is never meant to be returned to the storefront as is. The variants matrix will be used for filtering options allowed at the storefront after one or many options selected.
+
+### Variants, Prices & Inventory
+Variants could link the options intersection with a product, a price, or an inventory record but any of these relations is not mandatory.
+So, a variant can request data from the different domains if such data assigned.
+Such a decision brings us unseen before the level of the flexibility, since there is no difference between product price and variant price, and as the consequence variant price, could be included in B2B pricing.
+
+We do not have such behaviors either Magento 1 or 2 and as a result, the whole layer of product types such as bundle and downloadable do not support B2B prices or special prices for options.
+
+The link on a product that exists in the same domain could be done through the shared ID, which as for me makes sense since I can see a possible scenario when variation that initially had only a price will be "promoted" to a product. For example due to integration reasons, to pass additional attributes data along with variant to Google Merchant Center.
+
+### Enhanced option values
+
+To reach a better user experience,
+the option could be extended with an image that represents it 
+or info URL that provides additional information about option value.
+
+[Magento: Swatches](https://docs.magento.com/user-guide/catalog/swatches.html)
+
+[Magento: Downloadable products, options with samples](https://docs.magento.com/user-guide/catalog/product-create-downloadable.html)
```

Comment:
How can we cover the following attributes required for Downloadable product?
- sample URL, Link URL ("imageUrl" can cover only 1 url or we going to create 2 optionValues for "Downloadable Link")
- Shareable 
- Max. Downloads

Reply:
shareable and max download represent admin functionality and should not be accessible at the storefront. 
Samples are not options! you can not select a sample at the checkout, the closest by the meaning for samples are images and videos, so I expect that samples will be managed correspondingly.